norduser/service/fork.go

```diff
@@ -3,57 +3,135 @@ package service
+	return &ChildProcessNorduser{}
+}
+
+// handlePsError returns nil if err is nil or error code 1(no processes listed). It returns unmodified err in any other
+// case.
+func handlePsError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var exiterr *exec.ExitError
+	if errors.As(err, &exiterr) {
+		// ps returns 1 when no processes are shown
+		if exiterr.ExitCode() == 1 {
+			return nil
+		}
 	}
+
+	return err
 }
 
-func isRunning(uid uint32) (bool, error) {
-	// list all norduserd processes, restrict output to uid of the owner
+func parseNorduserPIDs(psOutput string) []int {
+	pids := []int{}
+	for _, pidStr := range strings.Split(psOutput, "\n") {
+		pid, err := strconv.Atoi(strings.TrimSpace(pidStr))
+		if err != nil {
+			log.Println("failed to parse pid string: ", err)
+			continue
+		}
+
+		pids = append(pids, pid)
+	}
+
+	return pids
+}
+
+func getRunningNorduserPIDs() ([]int, error) {
 	// #nosec G204 -- arguments are constant
-	output, err := exec.Command("ps", "-C", internal.Norduserd, "-o", "uid=").CombinedOutput()
-	if err != nil {
-		var exiterr *exec.ExitError
-		if errors.As(err, &exiterr) {
-			// ps returns 1 when no processes are shown
-			if exiterr.ExitCode() == 1 {
-				return false, nil
-			}
+	output, err := exec.Command("ps", "-C", internal.Norduserd, "-o", "pid=").CombinedOutput()
+	if err := handlePsError(err); err != nil {
+		return []int{}, fmt.Errorf("listing norduser pids: %w", err)
+	}
+
+	return parseNorduserPIDs(string(output)), nil
+}
+
+func findPIDOfUID(uids string, uid uint32) int {
+	desiredUID := fmt.Sprint(uid)
+	for _, pidUid := range strings.Split(uids, "\n") {
+		pidUidSplit := strings.Split(strings.TrimSpace(pidUid), " ")
+		if len(pidUidSplit) != 2 {
+			log.Println("unexpected ps output: ", pidUid)
 		}
 
-		return false, fmt.Errorf("listing processes: %w", err)
+		uid := pidUidSplit[0]
+		if uid != desiredUID {
+			continue
+		}
+
+		pid := pidUidSplit[1]
+		pidInt, err := strconv.Atoi(pid)
+		if err != nil {
+			log.Println("failed to parse pid: ", err)
+			continue
+		}
+
+		return pidInt
 	}
 
+	return -1
+}
+
+func getPIDForNorduserUID(uid uint32) (int, error) {
+	// list all norduserd processes, restrict output to uid of the owner
+	// #nosec G204 -- arguments are constant
+	output, err := exec.Command("ps", "-C", internal.Norduserd, "-o", "uid=", "-o", "pid=").CombinedOutput()
+	if err := handlePsError(err); err != nil {
+		return -1, fmt.Errorf("listing norduser uids/pids: %w", err)
+	}
+
+	return findPIDOfUID(string(output), uid), nil
+}
+
+func isUIDPresent(uids string, uid uint32) bool {
 	desiredUID := fmt.Sprint(uid)
-	uids := string(output)
 	for _, uid := range strings.Split(uids, "\n") {
 		if strings.Trim(uid, " ") == desiredUID {
-			return true, nil
+			return true
 		}
 	}
 
-	return false, nil
+	return false
+}
+
+func isRunning(uid uint32) (bool, error) {
+	// list all norduserd processes, restrict output to uid of the owner
```

Comment:
why not call `getPIDForNorduserUID` in this case and return true if `PID != -1`? 
might be less code to maintain :)